File: src/features/dashboard/graphEvents.ts
import { listen } from "@tauri-apps/api/event";

export interface GraphUpdateEvent {
  kind: "NodeCreated" | "NodeStateChanged" | "NodeArchived" | "EdgeCreated";
  id?: string;
  node_kind?: string;
  label?: string;
  parent_id?: string | null;
  detail?: string;
  tone?: string;
  source_id?: string;
  target_id?: string;
  edge_kind?: string;
}

export interface NotaDialogEvent {
  dialog_id: string;
  kind: "ApprovalRequired" | "Escalation" | "BudgetWarning" | "Info";
  title: string;
  body: string;
  context_json: string;
  allocation_id: number | null;
  transaction_id: number | null;
  actions: { action_key: string; label: string; tone: string }[];
}

export const listenToGraphUpdates = (callback: (event: GraphUpdateEvent) => void) =>
  listen<string>("graph:update", (event) => {
    try {
      callback(JSON.parse(event.payload) as GraphUpdateEvent);
    } catch {
      // Ignore malformed event payloads until backend emitters are fully wired.
    }
  });

export const listenToNotaDialogs = (callback: (event: NotaDialogEvent) => void) =>
  listen<string>("nota:dialog", (event) => {
    try {
      callback(JSON.parse(event.payload) as NotaDialogEvent);
    } catch {
      // Ignore malformed event payloads until backend emitters are fully wired.
    }
  });

File: src-tauri/src/core/graph_events.rs
use serde::Serialize;

#[derive(Debug, Clone, Serialize)]
#[serde(tag = "kind")]
pub enum GraphUpdateEvent {
    NodeCreated {
        id: String,
        node_kind: GraphNodeKind,
        label: String,
        parent_id: Option<String>,
        detail: String,
        tone: String,
    },
    NodeStateChanged {
        id: String,
        tone: String,
        detail: String,
    },
    NodeArchived {
        id: String,
    },
    EdgeCreated {
        source_id: String,
        target_id: String,
        edge_kind: String,
    },
}

#[derive(Debug, Clone, Serialize)]
pub enum GraphNodeKind {
    Nota,
    Allocation,
    Receipt,
    Checkpoint,
    Supervision,
    Dialog,
}

#[derive(Debug, Clone, Serialize)]
pub struct NotaDialogEvent {
    pub dialog_id: String,
    pub kind: NotaDialogKind,
    pub title: String,
    pub body: String,
    pub context_json: String,
    pub allocation_id: Option<i64>,
    pub transaction_id: Option<i64>,
    pub actions: Vec<DialogAction>,
}

#[derive(Debug, Clone, Serialize)]
pub enum NotaDialogKind {
    ApprovalRequired,
    Escalation,
    BudgetWarning,
    Info,
}

#[derive(Debug, Clone, Serialize)]
pub struct DialogAction {
    pub action_key: String,
    pub label: String,
    pub tone: String,
}
